README.md
# DigiOnline
An unofficial client addon for the Digi Online channels

## Disclaimer

Unofficial Kodi addon, not in any ways affiliated by DIGI Távközlési  és  Szolgáltató  Kft. or the "DigiOnline" service. Made only for educational purposes.

The author cannot take any responsibility for any possible damages.


plugin.video.dinionline/changelog.txt
[B]0.7.2[/B]
 - Matrix compatibility

[B]0.7.1[/B]
 - basic poller implementation so the provider won't kill our on-going streams during the playback

[B]0.7.0[/B]
 - Initial beta release


plugin.video.dinionline/utils/poller.py
# -*- coding: utf-8 -*-
import sys
import threading
import re
import xbmc

from mrdini.routines.routines import decrypt_string, request_page

NOT_PLAYING_WAIT_TIME = 0.2
PLAYING_WAIT_TIME = 5 * 60  # wait 5 minutes between each poll event
MAIN_URL = (
    "470098bXNyZXBvIGh0dHBzOi8vZGlnaW9ubGluZS5odQ=="
)  # TODO: better organize this
POLLER_URL_PREFIX = "%s/refresh" % (decrypt_string(MAIN_URL))


class PlaybackMonitorThread(threading.Thread):
    def __init__(self, utils):
        super(PlaybackMonitorThread, self).__init__()

        self._stopped = threading.Event()
        self._ended = (
            threading.Event()
        )  # probably not useful in livestreams' case but who knows
        self.utils = utils
        self.player = xbmc.Player()

        self.daemon = True
        self.start()

    def run(self):
        utils = self.utils
        before_playing_wait = 0.0
        while not self.player.isPlaying():
            xbmc.sleep(int(NOT_PLAYING_WAIT_TIME * 1000))
            if before_playing_wait >= 5:
                self.end()
                return  # couldn't start playing in 5 seconds
            before_playing_wait += NOT_PLAYING_WAIT_TIME
        while self.player.isPlaying():
            xbmc.log("Lejátszás folyamatban...")
            try:
                if (
                    not decrypt_string("470098bXNyZXBvIGh0dHBzOi8vb25saW5lLmRpZ2kuaHU=")
                    in self.player.getPlayingFile()
                ):
                    xbmc.log("Nem a mienk...")
                    self.stop()
                    break
                try:
                    c_id = (
                        self.player.getPlayingFile()
                        .split("playlist/", 1)[1]
                        .split("/", 1)[0]
                    )
                except:
                    xbmc.log("Sikertelen csatornanév kinyerés", xbmc.LOGERROR)
                    self.stop()
                    break
                headers = {
                    "Accept-Encoding": "gzip, deflate, br",
                    "Accept-Language": "hu-HU,hu;q=0.9,en-US;q=0.8,en;q=0.7",
                    "Accept": "*/*",
                    "Referer": "%s/player/%s" % (decrypt_string(MAIN_URL), c_id),
                    "X-Requested-With": "XMLHttpRequest",
                    "Connection": "keep-alive",
                }
                cookies = {
                    "XSRF-TOKEN": utils.get_setting("xsrf_token"),
                    "laravel_session": utils.get_setting("laravel_session"),
                    "acc_pp": utils.get_setting("acc_pp"),
                }
                response = request_page(
                    POLLER_URL_PREFIX,
                    params={"id": c_id},
                    cookies=cookies,
                    headers=headers,
                )
                xbmc.log(
                    "%s [%s] %s"
                    % (response.url, response.status_code, response.content)
                )
                for k, v in response.cookies.items():
                    if k == "XSRF-TOKEN":
                        utils.set_setting("xsrf_token", v)
                    elif k == "laravel_session":
                        utils.set_setting("laravel_session", v)
                    elif k == "acc_pp":
                        utils.set_setting("acc_pp", v)
                xbmc.log("Kérés elküldve...")
            except RuntimeError:
                pass
            xbmc.sleep(int(PLAYING_WAIT_TIME * 1000))
        self.end()

    def stop(self):
        self._stopped.set()

    def stopped(self):
        return self._stopped.is_set()

    def end(self):
        self._ended.set()

    def ended(self):
        return self._ended.is_set()


class DigiPlayer(xbmc.Player):
    def __init__(self, *args, **kwargs):
        self.utils = kwargs.get("utils")
        self.threads = []

    def stop_threads(self):
        for thread in self.threads:
            if thread.ended():
                continue

            if not thread.stopped():
                thread.stop()

        for thread in self.threads:
            if thread.stopped() and not thread.ended():
                try:
                    thread.join()
                except RuntimeError:
                    pass

    def cleanup_threads(self, only_ended=True):
        active_threads = []
        for thread in self.threads:
            if only_ended and not thread.ended():
                active_threads.append(thread)
                continue

            if not thread.ended() and not thread.stopped():
                thread.stop()
            try:
                thread.join()
            except RuntimeError:
                pass
        self.threads = active_threads

    def onPlayBackStarted(self):
        self.cleanup_threads()
        self.threads.append(PlaybackMonitorThread(self.utils))

    def onPlayBackEnded(self):
        self.stop_threads()
        self.cleanup_threads()

    def onPlayBackStopped(self):
        self.onPlayBackEnded()

    def onPlayBackError(self):
        self.onPlayBackEnded()


plugin.video.dinionline/utils/information.py
# -*- coding: utf-8 -*-
"""
    DigiOnline Kodi addon
    Copyright (C) 2019 Mr Dini
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.
    You should have received a copy of the GNU Affero General Public License
    along with this program. If not, see <http://www.gnu.org/licenses/>
"""
text = """%s - v%s\n
[B]Nem hivatalos kiegészítő, használat kizárólag saját felelősségre![/B] A használatból adódó esetleges következményekért a fejlesztő nem vállal felelősséget.
A kiegészítő béta állapotú, egyes funkciók hiányozhatnak.
Hibákat bátran jelenteni, esetleges módosításokat eszközölni a következő platformokon lehetséges:
Kiegészítő githubja: https://github.com/movieshark/DigiOnline
Fórumtéma: https://prohardver.hu/tema/kodi_xbmc_kiegeszito_magyar_nyelvu_online_filmekhe/friss.html
[B]Ikonok, háttér:[/B]\n
Icons made by Freepik from www.flaticon.com
[I]- By Dini[/I]
"""

eula = "Az Elfogadom gombra kattintva alulírom, hogy megértettem és elfogadom a szolgáltató adatvédelmi szabályzatát: [CR]%s"


plugin.video.dinionline/default.py
# -*- coding: utf-8 -*-
"""
    DigiOnline Kodi addon
    Copyright (C) 2019 Mr Dini
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.
    You should have received a copy of the GNU Affero General Public License
    along with this program. If not, see <http://www.gnu.org/licenses/>
"""

import sys
from re import findall

if sys.version_info[0] == 3:
    from urllib.parse import parse_qsl
else:
    # python2 compatibility
    from urlparse import parse_qsl
from mrdini.routines import routines, parsedom
from xbmcaddon import Addon
from xbmcplugin import endOfDirectory, setContent

utils = routines.Utils(Addon())

MAIN_URL = "470098bXNyZXBvIGh0dHBzOi8vZGlnaW9ubGluZS5odQ=="
EULA_URL = "470098bXNyZXBvIGh0dHA6Ly9kaWdpLmh1L2luZm9ybWFjaW8vYWRhdHZlZGVsbWktZXMtZmVsaGFzem5hbGFzaS1mZWx0ZXRlbGVr"


def update_cookies(response):
    token = parsedom.parseDOM(
        response.content, name="meta", attrs={"name": "csrf-token"}, ret="content"
    )
    if token:
        utils.set_setting("csrf_token", routines.py2_encode(token[0]))
    for k, v in response.cookies.items():
        if k == "XSRF-TOKEN":
            utils.set_setting("xsrf_token", v)
        elif k == "laravel_session":
            utils.set_setting("laravel_session", v)
        elif k == "acc_pp":
            utils.set_setting("acc_pp", v)


def cookie_builder():
    cookies = {
        "XSRF-TOKEN": utils.get_setting("xsrf_token"),
        "laravel_session": utils.get_setting("laravel_session"),
    }
    if utils.get_setting("acc_pp"):
        cookies["acc_pp"] = utils.get_setting("acc_pp")
    return cookies


def login():
    headers = {
        "Origin": routines.decrypt_string(MAIN_URL),
        "Accept-Encoding": "gzip, deflate, br",
        "X-CSRF-TOKEN": utils.get_setting("csrf_token"),
        "Accept-Language": "hu-HU,hu;q=0.9,en-US;q=0.8,en;q=0.7",
        "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
        "Accept": "*/*",
        "Referer": "%s/csatornak" % (routines.decrypt_string(MAIN_URL)),
        "X-Requested-With": "XMLHttpRequest",
        "Connection": "keep-alive",
    }
    data = {
        "email": utils.get_setting("username"),
        "password": utils.get_setting("password"),
        "accept": 1,  # the privacy rules
    }
    response = routines.request_page(
        "%s/login" % (routines.decrypt_string(MAIN_URL)),
        headers=headers,
        user_agent=utils.get_setting("user_agent"),
        cookies=cookie_builder(),
        data=data,
        allow_redirects=False,
    )
    update_cookies(response)
    if response.status_code != 302:
        utils.create_ok_dialog(
            "Helytelen bejelentkezési adatok! (Kód: %i)" % (response.status_code)
        )
        utils.open_settings()
        exit()
    utils.create_notification("Sikeres bejelentkezés!")


def main_window():
    routines.add_item(
        *sys.argv[:2],
        name="Élő Csatornakínálat",
        action="channels",
        is_directory=True,
        fanart=utils.fanart,
        icon="https://i.imgur.com/n0AbCQn.png"
    )
    routines.add_item(
        *sys.argv[:2],
        name="Mentett sütik törlése",
        action="clear_creds",
        description="A kiegészítő tárolja a sütiket, user-agentet és az azonosításhoz használatos egyedi tokent annak érdekében, hogy egy"
        " böngésző benyomását keltse, illetve spóroljon az adatforgalommal. Ez az opció hasznos lehet, amennyiben a bejelentkezési adatokat"
        " frissítetted, illetve ha szeretnél egy új user-agentet kapni a kiegészítőtől. Az adatok frissítése a legközelebbi csatornalista"
        " megnyitáskor történik.",
        is_directory=False,
        fanart=utils.fanart,
        icon="https://i.imgur.com/1T2DUvG.png"
    )
    routines.add_item(
        *sys.argv[:2],
        name="Beállítások",
        description="Addon beállításai",
        action="settings",
        is_directory=False,
        fanart=utils.fanart,
        icon="https://i.imgur.com/MI42pRz.png"
    )
    routines.add_item(
        *sys.argv[:2],
        name="A kiegészítőről",
        description="Egyéb infók",
        action="about",
        is_directory=False,
        fanart=utils.fanart,
        icon="https://i.imgur.com/bKJK0nc.png"
    )


def live_window():
    if not utils.get_setting("user_agent"):
        utils.set_setting("user_agent", routines.random_uagent())
    headers = {
        "Connection": "keep-alive",
        "Upgrade-Insecure-Requests": "1",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3",
        "Accept-Encoding": "gzip, deflate, br",
        "Accept-Language": "hu-HU,hu;q=0.9,en-US;q=0.8,en;q=0.7",
    }
    response = routines.request_page(
        "%s/csatornak" % (routines.decrypt_string(MAIN_URL)),
        user_agent=utils.get_setting("user_agent"),
        headers=headers,
    )
    update_cookies(response)
    login()
    for channel in parsedom.parseDOM(
        response.content, name="div", attrs={"class": '[^"]+channel'}
    ):
        try:
            url = "%s/player/%s" % (
                routines.decrypt_string(MAIN_URL),
                parsedom.parseDOM(
                    channel, "a", attrs={"data-id": '[^"]+'}, ret="data-id"
                )[0],
            )
        except:
            continue
            # if we have no way to stream, why would we even bother showing it
        try:
            logo = parsedom.parseDOM(
                parsedom.parseDOM(
                    channel, name="div", attrs={"class": ".+?channels__logo"}
                )[0],
                name="img",
                ret="src",
            )[0]
        except:
            logo = ""
        try:
            program_name = routines.py2_encode(
                parsedom.replaceHTMLCodes(
                    parsedom.parseDOM(
                        parsedom.parseDOM(
                            channel,
                            name="div",
                            attrs={"class": ".+?channels__program_name"},
                        )[0],
                        name="a",
                    )[0]
                )
            ).strip()
        except:
            program_name = ""
        try:
            name = routines.py2_encode(
                parsedom.replaceHTMLCodes(
                    parsedom.parseDOM(
                        channel, name="div", attrs={"class": "channels__name.+?"}
                    )[0]
                )
            ).strip()
        except:
            name = ""
        description = []
        try:
            timing = routines.py2_encode(
                parsedom.removeHTMLCodes(
                    parsedom.replaceHTMLCodes(
                        parsedom.parseDOM(
                            channel, name="p", attrs={"class": "timing.+?"}
                        )[0]
                    )
                )
            ).strip()
        except:
            timing = None
        try:
            progress = parsedom.parseDOM(
                channel,
                name="div",
                attrs={"class": "progress-bar"},
                ret="aria-valuenow",
            )[0].split(".", 1)[0]
        except:
            pass
        if timing and progress:
            description.append(
                "Sugárzási idő: %s (Jelenleg: %s%%)"
                % (routines.py2_encode(timing), routines.py2_encode(progress))
            )
        elif timing:
            description.append("Sugárzási idő: %s" % (routines.py2_encode(timing)))
        else:
            description.append("Jelenleg: %s%%" % (routines.py2_encode(progress)))
        try:
            description.append(
                routines.py2_encode(
                    parsedom.replaceHTMLCodes(
                        parsedom.parseDOM(
                            channel, name="p", attrs={"class": "description"}
                        )[0]
                    )
                ).strip()
            )
        except:
            pass
        try:
            description.append(
                "[COLOR yellow]Korhatár besorolás: %s[/COLOR]"
                % (
                    routines.py2_encode(
                        parsedom.parseDOM(
                            channel,
                            name="span",
                            attrs={"class": 'age-[^"]+'},
                            ret="class",
                        )[0]
                    ).replace("age-", "")
                )
            )
        except:
            pass
        try:
            description.append(
                "[I]%s[/I]"
                % (
                    routines.py2_encode(
                        parsedom.replaceHTMLCodes(
                            parsedom.parseDOM(
                                channel, name="div", attrs={"class": "next_program"}
                            )[0]
                        )
                    ).strip()
                )
            )
        except:
            pass
        if program_name and name:
            final_name = "%s[CR][COLOR gray]%s[/COLOR]" % (name, program_name)
        else:
            final_name = name or program_name
        routines.add_item(
            *sys.argv[:2],
            name=final_name,
            icon=logo,
            action="play",
            extra=url,
            description="[CR][CR]".join(description),
            fanart=utils.fanart,
            is_directory=False,
            refresh=True
        )

    setContent(int(sys.argv[1]), "tvshows")


def resolve_url(name, icon, url, description):
    headers = {
        "Connection": "keep-alive",
        "Upgrade-Insecure-Requests": "1",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3",
        "Accept-Encoding": "gzip, deflate, br",
        "Accept-Language": "hu-HU,hu;q=0.9,en-US;q=0.8,en;q=0.7",
    }
    response = routines.request_page(
        url,
        headers=headers,
        user_agent=utils.get_setting("user_agent"),
        cookies=cookie_builder(),
        allow_redirects=False,
    )
    update_cookies(response)
    url = findall(
        r"createDefaultPlayer\('([^']+)'[^\)]+\);",
        routines.py2_encode(response.content.decode("utf-8")),
    )
    if response.status_code == 302:
        utils.create_ok_dialog(
            "Nem sikerült a lejátszás. Ez a hiba akkor szokott előfordulni, ha párhuzamosan"
            " történik egy másik bejelentkezés. Kérem frissítse a csatornalistát az újbóli bejelentkezéshez!"
        )
        exit(1)
    elif not url or response.status_code != 200:
        utils.create_ok_dialog(
            "Nem sikerült kinyerni a lejátszási linket. Elképzelhető, hogy magasabb előfizetés szükséges."
        )
        exit()

    routines.play(
        int(sys.argv[1]),
        url[0],
        "video",
        user_agent=utils.get_setting("user_agent"),
        name=name,
        description=description,
        icon=icon,
    )


if __name__ == "__main__":
    params = dict(parse_qsl(sys.argv[2].replace("?", "")))
    action = params.get("action")
    if action is None:
        if utils.get_setting("is_firstrun") == "true":
            from utils.information import text, eula

            utils.create_textbox(text % (utils.addon_name, utils.version))
            if not utils.create_yesno_dialog(
                eula % (routines.decrypt_string(EULA_URL)),
                yes="Elfogadom",
                no="Kilépek",
            ):
                exit()
            utils.set_setting("is_firstrun", "false")
            utils.create_ok_dialog("Kérlek jelentkezz be az addon használatához!")
            utils.open_settings()
            exit()
        main_window()
        endOfDirectory(int(sys.argv[1]))
    if action == "channels":
        live_window()
        endOfDirectory(int(sys.argv[1]))
    if action == "clear_creds":
        utils.set_setting("xsrf_token", "")
        utils.set_setting("csrf_token", "")
        utils.set_setting("laravel_session", "")
        utils.set_setting("acc_pp", "")
        utils.set_setting("user_agent", "")
        utils.create_notification("Adatok sikeresen törölve!")
    if action == "play":
        resolve_url(
            params.get("name"),
            params.get("icon"),
            params.get("extra"),
            params.get("descr"),
        )
    if action == "settings":
        utils.open_settings()
    if action == "about":
        from utils.information import text

        utils.create_textbox(text % (utils.addon_name, utils.version))
